text/shared/src/main/scala/fs2/data/text/render/internal/StreamPrinter.scala

```diff
@@ -21,246 +21,275 @@ import cats.collections.Dequeue
 import cats.data.{Chain, NonEmptyList}
 import fs2.{Chunk, Pipe, Pull, Stream}
 
+private case class OpenGroup(hpl: Int, indent: Int, group: Chain[Annotated])
+private class AnnotationContext(var pos: Int,
+                                var aligns: NonEmptyIntList,
+                                var hpl: Int,
+                                var indent: Int,
+                                var groups: Dequeue[OpenGroup])
+private class RenderingContext(var fit: Int, var hpl: Int, var lines: NonEmptyList[String], var col: Int)
+
 private[render] class StreamPrinter[F[_], Event](width: Int, indentSize: Int)(implicit render: Renderable[Event])
     extends Pipe[F, Event, String] {
 
-  private val emptyGroups = (0, 0, Dequeue.empty[(Int, Int, Chain[Annotated])])
-
-  private def push(groups: Dequeue[(Int, Int, Chain[Annotated])],
-                   evt: Annotated): Dequeue[(Int, Int, Chain[Annotated])] =
-    groups.unsnoc match {
-      case Some(((ghpl, gindent, group), groups)) => groups.snoc((ghpl, gindent, group.append(evt)))
-      case None                                   => Dequeue.empty // should never happen
+  private def push(annctx: AnnotationContext, evt: Annotated): Unit =
+    annctx.groups.unsnoc match {
+      case Some((OpenGroup(ghpl, gindent, group), groups)) =>
+        annctx.groups = groups.snoc(OpenGroup(ghpl, gindent, group.append(evt)))
+      case None => // should never happen
```

Comment:
Do we have a better option than silently ignoring this bug? fs2 itself uses `assert()` to catch some bugs – not cool because AFAICT here we would still produce a semi-valid result, but let's at least have the discussion (and maybe a decision for all of fs2-data).

Reply:
Yeah, I don't like cases like this one, where we cannot statically ensure never happen. And I don't like crashing with an `assert` either. I will try to find something better here. But it was already like that before 😬